## Device initialization

This section initializes the computational environment for the project. It performs three critical tasks:

1. Detects available hardware (CPU vs GPU)
2. Configures PyTorch to use the optimal compute device
3. Verifies the setup by displaying device information

In [ ]:
import pandas as pd # Data manipulation and analysis.
import numpy as np # Numerical operations and array handling.
import matplotlib.pyplot as plt # More control, lower-level, basic plotting.
import seaborn as sns # Higher-level, more aesthetically pleasing plots.
from scipy import stats # Statistical functions and tests.

pd.set_option('display.max_columns', None) # Display all columns in DataFrame output.
pd.set_option('display.max_rows', None) # Display all rows in DataFrame output.

import torch
import types
import pandas as pd
import re
from collections import Counter
from torch.utils.data import DataLoader, Dataset
from torch import nn, optim
import copy

# Check if CUDA (GPU) is available for acceleration
print('is cuda available?', torch.cuda.is_available())

# If CUDA is available, print the name of the GPU device
if torch.cuda.is_available():
    print(torch.cuda.get_device_name(0))

# Set the device to GPU if available, otherwise use CPU
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print('Using device:', device)

is cuda available? False
Using device: cpu


## Dataset Loading

This section loads the SST dataset from local files and processes it for binary sentiment classification.

4 files are loaded from the `SST2-Data/stanfordSentimentTreebank/` folder:

### 1. `datasetSentences.txt`

Contains all sentences extracted from movie reviews, this is the primary data source containing the text we want to classify.

- Column 1: `sentence_index` - Unique identifier for each sentence
- Column 2: `sentence` - The actual text of the sentence

### 2. `datasetSplit.txt`

Specifies which dataset split each sentence belongs to, ensures using the standard train/dev/test splits.

- Column 1: `sentence_index` - Links to sentences in `datasetSentences.txt`
- Column 2: `splitset_label` - Split assignment
  - **1** = Training set
  - **2** = Test set
  - **3** = Development/Validation set

### 3. `dictionary.txt`

Maps all phrases (including sub-phrases) to unique IDs, acts as a lookup table to connect sentences to their sentiment labels, as a bridge between `datasetSentences.txt` and `sentiment_labels.txt`.

- Column 1: `phrase` - Text of the phrase (can be a word, sub-phrase, or complete sentence)
- Column 2: `phrase_id` - Unique integer identifier

### 4. `sentiment_labels.txt`

Contains sentiment scores for all phrases, provides the ground truth labels for training our sentiment classifier.

- Column 1: `phrase ids` - Links to `phrase_id` in `dictionary.txt`
- Column 2: `sentiment values` - Continuous sentiment score from 0 (most negative) to 1 (most positive)

There are 5 sentiments (very negative, negative, neutral, positive, very positive) in original dataset, but this project we only need binary classification (negative or positive), thus the sentences with sentiment values between 0.4 and 0.6 should be removed. Then convert labels to integers (0 or 1)

```
0.0 ←−−−−−−− 0.2 ←−−−− 0.4 ←− 0.5 −→ 0.6 −−−−→ 0.8 −−−−−−→ 1.0
Very Negative | Negative | Neutral | Positive | Very Positive
```
---

###  Final Format

After processing, each DataFrame has this structure:

| Column | Type | Description |
|--------|------|-------------|
| `sentence` | string | Movie review text |
| `label` | int | Binary sentiment |

In [2]:
# Load the four essential files from Stanford Sentiment Treebank dataset
# These files are interconnected and need to be merged to create our final dataset

# Load all sentences with their unique indices
sentences_df = pd.read_csv('SST2-Data/stanfordSentimentTreebank/stanfordSentimentTreebank/datasetSentences.txt', sep='\t', header=0)

# Load dataset split assignments (train/dev/test)
split_df = pd.read_csv('SST2-Data/stanfordSentimentTreebank/stanfordSentimentTreebank/datasetSplit.txt', sep=',', header=0)

# Load phrase dictionary (maps phrases to unique IDs)
dictionary_df = pd.read_csv('SST2-Data/stanfordSentimentTreebank/stanfordSentimentTreebank/dictionary.txt', sep='|', header=None, names=['phrase', 'phrase_id'])

# Load sentiment labels for all phrases
labels_df = pd.read_csv('SST2-Data/stanfordSentimentTreebank/stanfordSentimentTreebank/sentiment_labels.txt', sep='|', header=0)

print(f"Sentences: {len(sentences_df)}")           # Expected: 11,855
print(f"Dictionary phrases: {len(dictionary_df)}") # Expected: 239,232
print(f"Labels: {len(labels_df)}")                 # Expected: 239,232

# Combine sentences with their train/dev/test assignments
# This creates a unified DataFrame with: sentence_index, sentence, splitset_label
data = sentences_df.merge(split_df, on='sentence_index')

# Match Sentences to Sentiment Labels
# Process: Sentence Text → Dictionary (get phrase_id) → Labels (get sentiment) → Binary Label

# Initialize empty dictionary to store sentence-to-label mappings
sentence_labels = {}

# Iterate through each sentence in the merged dataset
for idx, row in data.iterrows():
    sentence = row['sentence']  # Extract the sentence text
    
    # Look up phrase_id in dictionary
    # Find the dictionary entry where phrase matches the current sentence
    phrase_match = dictionary_df[dictionary_df['phrase'] == sentence]
    
    # Check if we found a match (some sentences might not be in dictionary)
    if not phrase_match.empty:
        # Extract the phrase_id for this sentence
        phrase_id = phrase_match.iloc[0]['phrase_id']
        
        # Look up sentiment score using phrase_id
        # Find the label entry where phrase_id matches
        label_match = labels_df[labels_df['phrase ids'] == phrase_id]
        
        # Check if we found a sentiment score
        if not label_match.empty:
            # Extract the continuous sentiment value (0.0 to 1.0)
            sentiment_value = label_match.iloc[0]['sentiment values']
            
            # Convert continuous sentiment to binary label
            # Apply threshold-based conversion for binary classification
            
            if sentiment_value <= 0.4:
                # Negative class: very negative + negative samples
                # Range: [0.0, 0.4]
                sentence_labels[sentence] = 0
                
            elif sentiment_value >= 0.6:
                # Positive class: positive + very positive samples
                # Range: [0.6, 1.0]
                sentence_labels[sentence] = 1

# Apply Labels and Filter Neutral Samples
# Map the sentence_labels dictionary to create a new 'label' column
# Sentences not in sentence_labels (neutral samples) will have NaN values
data['label'] = data['sentence'].map(sentence_labels)

# Remove all rows where label is NaN
data = data.dropna(subset=['label'])

# Convert label column from float to integer type.
data['label'] = data['label'].astype(int)

# Display statistics after filtering
print(f"\nAfter filtering neutral samples: {len(data)}")  # Expected: ~9,613
print(f"Label distribution:\n{data['label'].value_counts()}")  # Should be ~50/50 split

# Split Data into Train/Validation/Test Sets
# Use the splitset_label column to separate data according to standard splits

# Training set (splitset_label == 1), keep only 'sentence' and 'label' columns, drop unnecessary columns
# reset_index(drop=True) renumbers rows from 0 to len-1 for clean indexing
train_df = data[data['splitset_label'] == 1][['sentence', 'label']].reset_index(drop=True)

# Test set (splitset_label == 2)
test_df = data[data['splitset_label'] == 2][['sentence', 'label']].reset_index(drop=True)

# Validation/Development set (splitset_label == 3)
val_df = data[data['splitset_label'] == 3][['sentence', 'label']].reset_index(drop=True)

# Display final dataset sizes
print(f"\nTrain: {len(train_df)}, Val: {len(val_df)}, Test: {len(test_df)}")

Sentences: 11855
Dictionary phrases: 239232
Labels: 239232

After filtering neutral samples: 9142
Label distribution:
label
1    4739
0    4403
Name: count, dtype: int64

Train: 6568, Val: 825, Test: 1749


## Dataset processing

This section transforms raw text data into numerical representations that neural networks can process. Text strings cannot be directly fed into neural networks - they must first be converted into sequences of integers.

### Tokenization

Split sentences into individual words and punctuation marks. Neural networks operate on discrete units. We also keep punctuation as separate tokens because they carry sentiment information

### Vocabulary Construction

Build a mapping between words and unique integer IDs (word → ID), a consistent way to convert any word into a number is required, including:

- Filter rare words (frequency < 2): Reduces vocabulary size and filters potential typos/noise
- Add special tokens: 
  - `<pad>`: For making sequences equal length (required for batch processing)
  - `<unk>`: For handling words not seen during training

In [3]:
# Tokenization: Breaking down text into individual words and punctuation marks
# This is the first step in converting raw text into numerical representations

_tok = re.compile(r"\w+|[^\w\s]")
def tokenize(s: str):
    """
        Convert a sentence string into a list of lowercase tokens.
        """
    return [t.lower() for t in _tok.findall(s)]

# Build Vocabulary from Training Data
# This allows to convert text (strings) into numbers that neural networks can process
# Define special tokens
UNK = "<unk>"  # Unknown token: used for words not in vocabulary (out-of-vocabulary words)
PAD = "<pad>"  # Padding token: used to make all sequences the same length

# Initialize a Counter to count word frequencies
# Counter is like a dictionary that automatically counts occurrences
counter = Counter()

# Count all words in the training set, only use training data to build vocabulary to prevent data leakage
for sentence in train_df['sentence']:
    # Tokenize each sentence and update word counts
    counter.update(tokenize(sentence))
# Build the vocabulary list (index-to-string mapping)
itos = [PAD, UNK] + [w for w, c in counter.most_common() if c >= 2]
 # Build reverse mapping: string-to-index dictionary, allows quick lookup of a word's ID
stoi = {w: i for i, w in enumerate(itos)}
# Get the padding token's index, will be used later to tell the model to ignore padding positions
padding_idx = stoi[PAD]  # padding_idx = 0
# Calculate vocabulary size
vocab_size = len(itos)

print(f"Vocabulary size: {vocab_size}")

Vocabulary size: 6864


## Loading Pre-trained Word Embeddings

This code loads pre-trained GloVe embeddings and integrates them with our vocabulary. Instead of learning word representations from scratch, we leverage embeddings trained on massive text corpora. Pre-trained embeddings significantly improve model performance, especially when training data is limited.

We now have a `TEXT` object containing our vocabulary mappings (`itos`, `stoi`) and an embedding matrix where known words start with meaningful representations, giving our model a significant advantage before training even begins.

In [4]:
# Import the gensim downloader API for accessing pre-trained word embeddings
import gensim.downloader as api

# Load pre-trained GloVe (Global Vectors for Word Representation) embeddings
glv = api.load("glove-wiki-gigaword-100")

# Set the dimensionality of word embeddings to match GloVe's dimension
embedding_dim = 100

# Initialize a tensor to store embedding vectors for all vocabulary words
pretrained_vectors = torch.randn(vocab_size, embedding_dim) * 0.01

# Iterate through each token in the vocabulary
for i, tok in enumerate(itos):
    # Check if the current token exists in the pre-trained GloVe vocabulary
    if tok in glv:
        # If found, replace the random vector with the pre-trained GloVe vector
        pretrained_vectors[i] = torch.tensor(glv[tok])

# Set the embedding vector for the padding token to zeros
pretrained_vectors[padding_idx] = 0.0

# Create a namespace object to mimic the torchtext Field structure
TEXT = types.SimpleNamespace(
    vocab=types.SimpleNamespace(
        itos=itos,                      # int-to-string: list mapping indices to tokens
        stoi=stoi,                      # string-to-int: dict mapping tokens to indices
        vectors=pretrained_vectors      # The embedding matrix with pre-trained vectors
    )
)

[==================================================] 100.0% 128.1/128.1MB downloaded


## Feature Engineering

This section implements a feature engineering pipeline to establish a baseline understanding of sentiment classification. First extract hand-crafted linguistic features and evaluate their predictive power. This approach helps us understand what signals are important for sentiment analysis.

There are 15 hand-crafted features extracted grouped into 7 categories, each capturing different aspects of sentiment expression:

1. Length Features (Text Complexity)

```python
- len_tokens: Number of word tokens
- len_chars: Total character count  
- avg_tok_len: Average token length
```

- Positive reviews might be longer (detailed praise) or negative reviews might be longer (detailed complaints)
- Longer words may indicate more formal or complex language
- Text length reflects how much the author cared to express

2. Punctuation Features (Emotional Expression)

People expressing strong sentiment use more emphatic punctuation.

```python
- count_exclaim: '!' count (excitement/emphasis)
- count_question: '?' count (uncertainty/questioning)
- count_period: '.' count (sentence structure)
- count_comma: ',' count (clause complexity)
- count_punct_total: Overall punctuation density
```

- Exclamation marks signal strong emotion (both positive and negative)
- Question marks often indicate confusion, doubt, or rhetorical emphasis
- Commas suggest detailed, structured arguments (common in thoughtful reviews)
- Punctuation density reflects writing style and emotional intensity

3. Uppercase Features (Emphasis/Shouting)

```python
- count_upper_tokens: Number of ALL CAPS words
- ratio_upper_tokens: Proportion of text in caps
```

- ALL CAPS = SHOUTING or EMPHASIS
- Strong indicator of emotional intensity

4. Elongation Features (Informal Emphasis)

Character elongation mimics prosody (speech intonation) in written text.

```python
- count_elongated: Words with repeated characters
```

- Informal emphasis technique common in social media and reviews
- Strong sentiment indicator: Positive: "This is soooo good!" Negative: "Soooo disappointed"

---

5. Negation Features (Sentiment Reversal)

Simple bag-of-words models fail on negations - this feature explicitly captures them.

```python
- count_negations: Negation words (not, no, never, don't, can't, etc.)
```

- Negations reverse polarity: "good" → positive, "not good" → negative
- Negations interact with surrounding words to flip meaning
- High negation count may indicate negative sentiment ("not good", "don't like", "never again")

6. Intensifier Features (Sentiment Amplification)

Intensifiers modify adjectives to increase subjective intensity.

```python
- count_intensifiers: Intensifying adverbs (very, really, extremely, incredibly)
```

- Amplify sentiment strength without changing polarity
- Indicate strong opinions and emotional engagement
- Presence suggests the author felt compelled to emphasize their sentiment

7. Emoticon Features (Direct Sentiment Signals)

Emoticons serve as digital paralanguage, conveying emotion that would be expressed through facial expressions in person.

```python
- count_pos_emotes: Positive emoticons (:), :D, ^^)
- count_neg_emotes: Negative emoticons (:(, :/, :'()
```

- Unambiguous sentiment indicators - direct emotional expression
- Common in informal text (social media, casual reviews)
- Often used to clarify tone: "This is great :)" vs. "This is great :/"

### Feature Normalization Strategy

All features are normalized to prevent scale imbalance:

- Raw character count (100s) would overwhelm emoticon count (0-2)
- A text with 100 tokens and 5 exclamations is different from 10 tokens and 5 exclamations

using flowing methods:

- Length features → Scaled by typical text size (20 tokens)
- Punctuation → Divided by token or character count
- Ratios → Already normalized (e.g., `ratio_upper_tokens`)
- Emoticons → Raw counts (rare events, no scaling needed)

### Chi-Square Feature Importance Test

The Chi-Square test measures the statistical dependency between each feature and the sentiment label. A high χ² score indicates a strong association between the feature and sentiment, suggesting the feature is informative for prediction. Additionally, a low p-value confirms that this relationship is statistically significant and not due to random chance. 

We can keep features with high χ² scores (strong predictors) and discard those with low scores (weak or redundant predictors), ensuring our model focuses on the most informative signals while reducing noise and dimensionality.

### Logistic Regression Baseline

Logistic Regression serves as a linear model, it learns explicit weights for each feature, allowing us to understand exactly how much each linguistic pattern contributes to sentiment prediction. This transparency is valuable for validating our feature engineering decisions and building intuition about what matters for sentiment classification. 

An accuracy of 70-75% indicates that our hand-crafted features successfully capture significant sentiment signals, validating our feature engineering approach while leaving room for more sophisticated models to improve. If accuracy falls between 50-60%, it suggests our features are too weak and that we need representation learning techniques (like word embeddings and neural networks) to extract deeper semantic patterns. Conversely, if accuracy exceeds 80%, it may indicate that the classification problem is relatively simple, the dataset is small or homogeneous

### Ablation Study

Ablation analysis is a systematic experimental approach where we remove one feature group at a time and measure how much the model's performance degrades. By training separate models with each feature group excluded, we can directly observe each group's contribution to overall accuracy. A large accuracy drop indicates the removed feature group is critical and should be retained, a small or negligible drop suggests the group is redundant or already captured by other features, and paradoxically, if accuracy improves after removal, it means the feature group was adding noise rather than signal and should definitely be eliminated.

Ablation analysis enables principled dimensionality reduction by revealing which features are truly unhelpful, allowing us to streamline our feature set and reduce noise without sacrificing predictive power. 

### Feature Selection Decision

Use statistical significance (chi-square) as primary filter, validate with ablation study.

1. Chi-square score > 10: Strong statistical association with sentiment
2. Ablation study confirms importance**: Removing causes accuracy drop
3. Interpretability: Feature makes linguistic sense

In [ ]:
from sklearn.metrics import accuracy_score          # Metric to evaluate classification accuracy
from sklearn.discriminant_analysis import StandardScaler  # Feature normalization (z-score scaling)
from sklearn.feature_selection import chi2       # Chi-square test for feature importance
from sklearn.linear_model import LogisticRegression  # Simple linear classifier as baseline

# Words that negate or reverse sentiment (e.g., "not good" becomes negative)
NEGATIONS = {"not","no","never","n't","dont","don't","didn't","won't","cannot","can't"}

# Words that amplify sentiment intensity (e.g., "very good" is stronger than "good")
INTENSIFIERS = {"very","really","so","too","extremely","super","highly","utterly","absolutely","incredibly"}

# Emoticons indicating positive sentiment
POS_EMOTES = [":)",":-)",":d","=)",":]","^^"]

# Emoticons indicating negative sentiment
NEG_EMOTES = [":("," :-("," ):",":'("," :/ ",":-/"]


# These features capture linguistic patterns that correlate with sentiment
NUMERIC_FEATURE_NAMES = [
    "len_tokens",           # Number of word tokens (vocabulary richness)
    "len_chars",            # Total character count (text verbosity)
    "avg_tok_len",          # Average token length (word complexity)
    
    "count_exclaim",        # '!' count (excitement/emphasis indicator)
    "count_question",       # '?' count (questioning/uncertainty)
    "count_period",         # '.' count (sentence structure)
    "count_comma",          # ',' count (clause complexity)
    "count_punct_total",    # Total punctuation (stylistic density)
    
    "count_upper_tokens",   # ALL CAPS words (shouting/emphasis)
    "ratio_upper_tokens",   # Proportion of uppercase tokens
    
    "count_elongated",      # Words with repeated letters (e.g., "soooo good")
    "count_negations",      # Negation words (sentiment reversal)
    "count_intensifiers",   # Intensifier words (sentiment amplification)
    
    "count_pos_emotes",     # Positive emoticon count
    "count_neg_emotes"      # Negative emoticon count
]


def is_all_caps_token(tok: str) -> bool:
    """
    Check if a token is written in ALL CAPS.
    """
    # Extract only alphabetic characters
    letters = [ch for ch in tok if ch.isalpha()]
    # Require at least 2 letters and all must be uppercase
    return (len(letters) >= 2) and all(ch.isupper() for ch in letters)


def compute_features_for_text(text: str):
    """
    Extract hand-crafted linguistic features from text for sentiment analysis.
    
    This function computes 15 normalized features that capture:
    - Text length and complexity
    - Punctuation patterns
    - Stylistic markers (caps, elongation)
    - Sentiment modifiers (negations, intensifiers)
    - Emotional indicators (emoticons)
    """
    # Tokenize text using pre-defined tokenizer pattern
    toks = _tok.findall(text)
    toks_lower = [t.lower() for t in toks]
    
    # Filter to word tokens (containing at least one letter)
    word_tokens = [t for t in toks if any(ch.isalpha() for ch in t)]
    T = len(word_tokens)  # Total number of word tokens
    L = len(text)         # Total character length
    
    # PUNCTUATION FEATURES: Capture emotional expression through punctuation
    count_exclaim = text.count("!")        # Excitement/emphasis
    count_question = text.count("?")       # Questions/uncertainty
    count_period = text.count(".")         # Sentence boundaries
    count_comma = text.count(",")          # Clause complexity
    # Total non-alphanumeric, non-space characters
    count_punct_total = sum(1 for ch in text if (not ch.isalnum()) and (not ch.isspace()))
    
    # STYLISTIC FEATURES: Capture writing style indicators
    count_upper_tokens = sum(1 for t in word_tokens if is_all_caps_token(t))
    ratio_upper_tokens = (count_upper_tokens / T) if T > 0 else 0.0
    
    # Elongated words indicate emphasis, any character repeated 3+ times
    count_elongated = sum(1 for t in toks_lower 
                         if any(ch.isalpha() for ch in t) and re.search(r"(.)\1{2,}", t))
    
    # SENTIMENT MODIFIER FEATURES
    count_negations = sum(1 for t in toks_lower if t in NEGATIONS) + text.lower().count("n't")
    
    # Intensifiers amplify sentiment strength
    count_intensifiers = sum(1 for t in toks_lower if t in INTENSIFIERS)
    
    # EMOTICON FEATURES: Direct sentiment indicators
    tl = text.lower()
    count_pos_emotes = sum(tl.count(e.strip()) for e in POS_EMOTES)
    count_neg_emotes = sum(tl.count(e.strip()) for e in NEG_EMOTES)
    
    # LENGTH FEATURES: Basic text statistics
    len_tokens = float(T)
    len_chars = float(L)
    avg_tok_len = (len_chars / len_tokens) if T > 0 else 0.0
    
    # NORMALIZATION: Scale features to comparable ranges
    T_norm = max(T, 1.0)
    
    # Normalize each feature to prevent any single feature from dominating
    feats = {
        # Length features: scale relative to text size
        "len_tokens_norm": len_tokens / max(T_norm, 20.0),
        "len_chars_norm": len_chars / (4.0 * T_norm),  # Assume ~4 chars per token
        "avg_tok_len_norm": avg_tok_len / 10.0,
        
        # Punctuation: normalize by token count
        "count_exclaim_norm": count_exclaim / T_norm,
        "count_question_norm": count_question / T_norm,
        "count_period_norm": count_period / T_norm,
        "count_comma_norm": count_comma / T_norm,
        "count_punct_total_norm": count_punct_total / max(L, 1.0),
        
        # Uppercase: scale by half token count (typically rare)
        "count_upper_tokens_norm": count_upper_tokens / max(T_norm/2.0, 1.0),
        "ratio_upper_tokens_norm": ratio_upper_tokens,  # Already a ratio
        
        # Style modifiers: normalize by token count
        "count_elongated_norm": count_elongated / T_norm,
        "count_negations_norm": count_negations / T_norm,
        "count_intensifiers_norm": count_intensifiers / T_norm,
        
        # Emoticons: raw counts (typically 0-2)
        "count_pos_emotes_norm": float(count_pos_emotes),
        "count_neg_emotes_norm": float(count_neg_emotes),
    }
    return feats


print("Building feature validation dataset...")
feat_rows = []

# Extract features for each training example
for idx, row in train_df.iterrows():
    feats = compute_features_for_text(row['sentence'])
    feats["label"] = row['label']  # Attach ground truth label
    feat_rows.append(feats)

# Convert to DataFrame for analysis
df_features = pd.DataFrame(feat_rows)

# Get list of normalized feature columns (exclude label)
feature_cols = [c for c in df_features.columns if c.endswith("_norm")]

print(f"Number of features: {len(feature_cols)}")
print(f"Training samples: {len(df_features)}")


# Prepare feature matrix and labels
X_chi = df_features[feature_cols].values
y_chi = df_features["label"].values

# Chi-square test measures dependency between each feature and the label
# Higher chi2 score = stronger association with sentiment
chi_vals, p_vals = chi2(X_chi, y_chi)

# Create ranking table
chi_df = pd.DataFrame({
    "feature": feature_cols,
    "chi2_score": chi_vals,      # Higher = more informative
    "p_value": p_vals            # Lower = more statistically significant
}).sort_values("chi2_score", ascending=False)

print("\nFeature Importance Ranking (Chi-square):")
print(chi_df.to_string(index=False))

# Extract features from validation set
val_feat_rows = []
for idx, row in val_df.iterrows():
    feats = compute_features_for_text(row['sentence'])
    feats["label"] = row['label']
    val_feat_rows.append(feats)

df_val = pd.DataFrame(val_feat_rows)

# Prepare train/val splits
X_train_lr = df_features[feature_cols].values
y_train_lr = df_features["label"].values
X_val_lr = df_val[feature_cols].values
y_val_lr = df_val["label"].values

# Standardize features (zero mean, unit variance)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_lr)
X_val_scaled = scaler.transform(X_val_lr)

# Train logistic regression classifier
# This establishes a baseline: how well can hand-crafted features perform?
lr_model = LogisticRegression(max_iter=1000, solver="liblinear")
lr_model.fit(X_train_scaled, y_train_lr)

# Evaluate on validation set
lr_pred = lr_model.predict(X_val_scaled)
lr_acc = accuracy_score(y_val_lr, lr_pred)

print(f"\nLR Baseline validation accuracy: {lr_acc:.4f}")
print(f"Conclusion: Hand-crafted features alone achieve {lr_acc*100:.1f}% classification ability")

# Group features by category for systematic analysis
feature_groups = {
    "length": ["len_tokens_norm","len_chars_norm","avg_tok_len_norm"],
    "punctuation": ["count_exclaim_norm","count_question_norm","count_period_norm",
                   "count_comma_norm","count_punct_total_norm"],
    "uppercase": ["count_upper_tokens_norm","ratio_upper_tokens_norm"],
    "elongated": ["count_elongated_norm"],
    "negations": ["count_negations_norm"],
    "intensifiers": ["count_intensifiers_norm"],
    "emoticons": ["count_pos_emotes_norm","count_neg_emotes_norm"],
}

ablation_results = []

# For each feature group, remove it and measure performance drop
# This reveals which features contribute most to classification
for group_name, group_feats in feature_groups.items():
    # Remove current group's features
    remaining_feats = [f for f in feature_cols if f not in group_feats]
    
    X_train_abl = df_features[remaining_feats].values
    X_val_abl = df_val[remaining_feats].values
    
    # Standardize remaining features
    scaler_abl = StandardScaler()
    X_train_abl_scaled = scaler_abl.fit_transform(X_train_abl)
    X_val_abl_scaled = scaler_abl.transform(X_val_abl)
    
    # Train and evaluate with reduced feature set
    lr_abl = LogisticRegression(max_iter=1000, solver="liblinear")
    lr_abl.fit(X_train_abl_scaled, y_train_lr)
    pred_abl = lr_abl.predict(X_val_abl_scaled)
    acc_abl = accuracy_score(y_val_lr, pred_abl)
    
    # Record performance drop when this group is removed
    ablation_results.append({
        "removed_group": group_name,
        "val_acc": acc_abl,
        "acc_drop": lr_acc - acc_abl  # Positive = removing hurts performance
    })

# Sort by accuracy drop to identify most important feature groups
ablation_df = pd.DataFrame(ablation_results).sort_values("acc_drop", ascending=False)
print("\nAblation Study Results (performance after removing feature groups):")
print(ablation_df.to_string(index=False))

# Select features with chi-square score above threshold
# This filters out weakly predictive features
IMPORTANCE_THRESHOLD = 10.0
selected_features = chi_df[chi_df["chi2_score"] > IMPORTANCE_THRESHOLD]["feature"].tolist()

print(f"Based on chi-square > {IMPORTANCE_THRESHOLD}, selected {len(selected_features)} features:")
for feat in selected_features:
    chi_val = chi_df[chi_df["feature"] == feat]["chi2_score"].values[0]
    print(f"   - {feat:<30} (chi2={chi_val:.2f})")

Building feature validation dataset...
Number of features: 15
Training samples: 6568

Feature Importance Ranking (Chi-square):
                feature  chi2_score  p_value
   count_negations_norm   11.013588 0.000904
count_intensifiers_norm    5.073290 0.024297
    count_question_norm    4.994573 0.025427
     count_exclaim_norm    3.474562 0.062319
         len_chars_norm    2.561395 0.109502
       count_comma_norm    2.267597 0.132105
      count_period_norm    1.148853 0.283789
       avg_tok_len_norm    1.024558 0.311440
count_upper_tokens_norm    1.018557 0.312862
ratio_upper_tokens_norm    0.509278 0.475451
 count_punct_total_norm    0.328244 0.566695
   count_elongated_norm    0.266407 0.605752
        len_tokens_norm    0.000239 0.987673
  count_pos_emotes_norm         NaN      NaN
  count_neg_emotes_norm         NaN      NaN

LR Baseline validation accuracy: 0.5952
Conclusion: Hand-crafted features alone achieve 59.5% classification ability

Ablation Study Results (performanc

## DataLoader Setup

Create efficient data pipelines that feed batches of samples to the model during training.

### Dataset Class

Wrap preprocessed data in PyTorch's Dataset interface. PyTorch's DataLoader requires this structure for efficient batching and data loading

This pipeline ensures our text data is properly prepared for the LSTM model while maintaining reproducibility and preventing data leakage.


### Compute Numeric Feature Function

The function processes batches of tokenized text in real-time, converting token IDs back to their string representations to analyze character-level patterns (punctuation, capitalization, emoticons) that aren't captured by word embeddings alone. These features are then normalized to comparable scales and concatenated with the neural network's learned representations, creating a richer input that combines learned semantic knowledge with explicit linguistic rules.

In the neural network forward pass, this function is called to compute features for each batch.

### Collate Function 

Sentences have variable lengths (e.g., 5 words vs 20 words), but neural networks require fixed-size tensors. To address this, pad shorter sequences with special `<pad>` tokens to match the longest sequence in each batch

### Training DataLoader
- `shuffle=True`: Randomize sample order each epoch to prevent overfitting to data order, which is critical for good generalization

- `batch_size=128`: Balance between speed and memory, 128 is a common choice for medium datasets

### Validation/Test DataLoaders
- `shuffle=False`: Keep fixed order to reproducible evaluation metrics

- `pin_memory=True`: Speed up CPU→GPU transfer. Only beneficial when using CUDA

This setup ensures our data flows efficiently from disk → CPU → GPU during training while maintaining reproducibility for evaluation.

In [ ]:
# Define Custom PyTorch Dataset Class for accessing individual samples during training
class SST2Dataset(Dataset):
    """
    Custom PyTorch Dataset for SST-2 sentiment classification.
    Converts text sentences into numerical representations (token IDs) and pairs them with their corresponding labels.
    """
    
    def __init__(self, dataframe, stoi):
        """
        Initialize the dataset by preprocessing all samples.
        """
        # Pre-process all data and store in memory
        self.data = []
        # Iterate through each row in the dataframe
        for _, row in dataframe.iterrows():
            sentence = row['sentence']
            label = int(row['label'])
            # Convert sentence to list of integer IDs
            ids = [stoi.get(w, stoi[UNK]) for w in tokenize(sentence)]
            # Store as dictionary for easy access
            self.data.append({'input_ids': ids, 'label': label})
    
    def __len__(self):
        """
        Return the total number of samples in the dataset.
        """
        return len(self.data)
    
    def __getitem__(self, idx):
        """
        Return a single sample at the given index.
        This is called by DataLoader to fetch individual samples.
        """
        return self.data[idx]

# Instantiate dataset objects for each split
# These objects will be used by DataLoader during training
train_ds = SST2Dataset(train_df, stoi)
val_ds = SST2Dataset(val_df, stoi)
test_ds = SST2Dataset(test_df, stoi)
# Display dataset sizes for verification
print(f"Dataset sizes - Train: {len(train_ds)}, Val: {len(val_ds)}, Test: {len(test_ds)}")


# Extract hand-crafted linguistic features from tokenized text batches.
def compute_numeric_features(x_ids, pad_idx, selected_feat_names):
    """
    This function converts token IDs back to text and computes numerical features
    that capture sentiment-related linguistic patterns. These features complement
    neural network representations by explicitly encoding domain knowledge about
    sentiment expression.
    """
    # Preserve the device (CPU/GPU) of input tensor to ensure output stays on same device
    device = x_ids.device
    
    # Get batch size (B) and sequence length (T)
    B, T = x_ids.size()
    
    # List to accumulate feature vectors for each sample in the batch
    feats_list = []
    
    # Process each text sample in the batch independently
    for i in range(B):
        # Extract token IDs for current sample and convert to Python list
        ids = x_ids[i].tolist()
        
        # Remove padding tokens - only analyze actual content
        ids = [t for t in ids if t != pad_idx]
        
        # Map token IDs back to string tokens using vocabulary
        toks = [TEXT.vocab.itos[t] for t in ids]
        
        # Reconstruct the original text by joining tokens with spaces
        text = " ".join(toks)
        
        # Compute basic text statistics
        len_tokens = float(len(toks))          # Number of word tokens
        len_chars = float(len(text))           # Total character count
        avg_tok_len = (len_chars / len_tokens) if len_tokens > 0 else 0.0  # Average word length
        
        # Count punctuation marks (emotional expression indicators)
        count_exclaim = float(text.count("!"))     # Exclamation marks (emphasis/excitement)
        count_question = float(text.count("?"))    # Question marks (uncertainty/inquiry)
        count_period = float(text.count("."))      # Periods (sentence structure)
        count_comma = float(text.count(","))       # Commas (clause complexity)
        
        # Count all non-alphanumeric, non-space characters (overall punctuation density)
        count_punct_total = float(sum(1 for ch in text if (not ch.isalnum()) and (not ch.isspace())))
        
        # Analyze capitalization patterns (emphasis/shouting)
        count_upper_tokens = float(sum(1 for tok in toks if is_all_caps_token(tok)))
        
        # Ratio of uppercase tokens to total tokens
        ratio_upper_tokens = (count_upper_tokens / len_tokens) if len_tokens > 0 else 0.0
        
        # Detect elongated words (informal emphasis)
        count_elongated = float(sum(1 for tok in toks if re.search(r"(.)\1{2,}", tok or "")))
        
        # Count sentiment modifiers
        toks_lower = [t.lower() for t in toks]
        
        # Count negation words (sentiment polarity reversers)
        count_negations = float(sum(1 for t in toks_lower if t in NEGATIONS))
        
        # Count intensifier words (sentiment amplifiers)
        count_intensifiers = float(sum(1 for t in toks_lower if t in INTENSIFIERS))
        
        # Count emoticons (direct sentiment signals)
        tl = text.lower()  # Lowercase for case-insensitive emoticon matching
        
        # Sum occurrences of positive emoticons
        count_pos_emotes = float(sum(tl.count(e.strip()) for e in POS_EMOTES))
        
        # Sum occurrences of negative emoticons
        count_neg_emotes = float(sum(tl.count(e.strip()) for e in NEG_EMOTES))
        
        # Store raw feature values
        all_feats = {
            "len_tokens": len_tokens,
            "len_chars": len_chars,
            "avg_tok_len": avg_tok_len,
            "count_exclaim": count_exclaim,
            "count_question": count_question,
            "count_period": count_period,
            "count_comma": count_comma,
            "count_punct_total": count_punct_total,
            "count_upper_tokens": count_upper_tokens,
            "ratio_upper_tokens": ratio_upper_tokens,
            "count_elongated": count_elongated,
            "count_negations": count_negations,
            "count_intensifiers": count_intensifiers,
            "count_pos_emotes": count_pos_emotes,
            "count_neg_emotes": count_neg_emotes,
        }
        
        # Use actual token count for normalization, with minimum of 1.0 to avoid division by zero
        T_norm = max(T, 1.0)
        
        all_feats_norm = {
            # Length features: normalize by typical text size
            "len_tokens_norm": all_feats["len_tokens"] / max(T_norm, 20.0),
            "len_chars_norm": all_feats["len_chars"] / (4.0 * T_norm),  # Assume ~4 chars/token
            "avg_tok_len_norm": all_feats["avg_tok_len"] / 10.0,         # Scale to 0-1 range
            
            # Punctuation features: normalize by token count (frequency per token)
            "count_exclaim_norm": all_feats["count_exclaim"] / T_norm,
            "count_question_norm": all_feats["count_question"] / T_norm,
            "count_period_norm": all_feats["count_period"] / T_norm,
            "count_comma_norm": all_feats["count_comma"] / T_norm,
            "count_punct_total_norm": all_feats["count_punct_total"] / max(len(text), 1.0),
            
            # Uppercase features: scale by half token count (uppercase is typically rare)
            "count_upper_tokens_norm": all_feats["count_upper_tokens"] / max(T_norm/2.0, 1.0),
            "ratio_upper_tokens_norm": all_feats["ratio_upper_tokens"],  # Already a ratio (0-1)
            
            # Style modifier features: normalize by token count
            "count_elongated_norm": all_feats["count_elongated"] / T_norm,
            "count_negations_norm": all_feats["count_negations"] / T_norm,
            "count_intensifiers_norm": all_feats["count_intensifiers"] / T_norm,
            
            # Emoticon features: use raw counts (typically 0-2, already small scale)
            "count_pos_emotes_norm": all_feats["count_pos_emotes"],
            "count_neg_emotes_norm": all_feats["count_neg_emotes"],
        }
        
        # Extract only the selected features
        selected_vals = [all_feats_norm[fname] for fname in selected_feat_names]
        
        # Add this sample's feature vector to the batch list
        feats_list.append(selected_vals)
    
    # Convert to tensor and return
    return torch.tensor(feats_list, dtype=torch.float32, device=device)


# Combines multiple samples into a single batch
# Neural networks require fixed-size inputs for efficient batch processing

def collate_fn(batch):
    """
    Custom collate function to process a batch of samples.
    """
    
    # Initialize lists to collect input sequences and labels
    ids_list = []
    labels = []
    
    # Extract data from each sample in the batch
    for item in batch:
        ids = item['input_ids']    # List of token IDs (variable length)
        label = item['label']       # Integer label (0 or 1)
        ids_list.append(ids)
        labels.append(label)
    
    # Find the maximum sequence length in this batch
    max_len = max(len(x) for x in ids_list)
    
    # Create a tensor filled with padding_idx (0)
    # All positions are initialized to padding_idx and will be overwritten with actual tokens
    x = torch.full((len(ids_list), max_len), padding_idx, dtype=torch.long)
    
    # Fill in the actual token IDs for each sequence
    for i, ids in enumerate(ids_list):
        # Copy token IDs to the beginning of each row
        x[i, :len(ids)] = torch.tensor(ids, dtype=torch.long)
    
    # Convert labels list to tensor
    y = torch.tensor(labels, dtype=torch.long)
    
    # Return as SimpleNamespace for convenient attribute access
    return types.SimpleNamespace(text=x, label=y)

# Create DataLoader Objects
# PyTorch utility that handles batching, shuffling, and parallel data loading

# Training DataLoader
train_iter = DataLoader(
    train_ds,              # Dataset object to load from
    batch_size=128,        # Number of samples per batch
    shuffle=True,          # Randomly shuffle data each epoch
    collate_fn=collate_fn, # Our custom function to combine samples into batches
    pin_memory=True        # If True, allocates tensors in pinned memory
)

# Validation DataLoader
val_iter = DataLoader(
    val_ds,                # Validation dataset
    batch_size=128,        # Same batch size as training for consistency
    shuffle=False,         # DON'T shuffle validation data
    collate_fn=collate_fn, # Same collate function
    pin_memory=True
)

# Test DataLoader
test_iter = DataLoader(
    test_ds,               # Test dataset
    batch_size=128,
    shuffle=False,         # DON'T shuffle test data
    collate_fn=collate_fn,
    pin_memory=True
)


# Get the first batch from training DataLoader
first = next(iter(train_iter))

# Display batch information for verification
print(f"\nBatch info:")
print(f"Type: {type(first)}")  
print(f"Text shape: {first.text.shape}")  
print(f"Label shape: {first.label.shape}")  
print(f"Sample text (first 10 tokens): {first.text[0][:10]}")  
print(f"Sample label: {first.label[0]}")

## BiLSTM Sentiment Classification Model

For sentiment classification, we employ a Bidirectional Long Short-Term Memory (BiLSTM) network as our primary algorithm. This choice is justified by several key advantages:

1. Sequential Nature of Language: Text is not just a collection of random words—the order matters tremendously. LSTMs are designed specifically to process sequences where order matters. As the LSTM reads through a sentence word by word, it maintains a "memory" of what it has seen before. 

2. Long-Range Dependency Modeling: Sometimes the key to understanding sentiment is remembering something from much earlier in the text. RNNs  try to do this but fail because they forget information from earlier in the sequence. LSTMs solve this problem using special mechanisms called "gates" that act like memory controllers. 

3. Bidirectional Context Capture: A regular LSTM only reads left-to-right (forward). A Bidirectional LSTM  reading the sentence in both directions simultaneously: one LSTM reads left-to-right (forward), and another reads right-to-left (backward). By combining information from both directions, the model gets the complete picture, leading to much better understanding of sentiment.

4. Hierarchical Representation Learning: We use 2 LSTM layers stacked on top of each other, which allows the model to learn patterns at different levels of abstraction. Our first LSTM layer learns basic patterns like grammar and simple word combinations. The second LSTM layer then takes these basic patterns and learns more complex, abstract patterns. 

### Metrics

#### 1. Classification Accuracy

We mainly use classification accuracy as our primary performance measure. 

Classification accuracy is the most appropriate metric for our sentiment analysis task. First, our dataset exhibits balanced class distribution—we have roughly equal numbers of positive and negative reviews, meaning accuracy won't be misleadingly inflated by predicting the majority class.

Second, for binary sentiment classification, both classes (positive/negative) are equally important. We don't have an asymmetric cost structure where false positives are more costly than false negatives, so accuracy's equal weighting of all errors is appropriate.

#### 2. Cross-Entropy Loss

We also monitor cross-entropy loss during training and validation.

Cross-entropy loss measures the divergence between predicted probability distributions and true labels, providing a more nuanced signal than accuracy during training. Cross-entropy rewards confident correct predictions and penalizes confident wrong predictions more heavily. 

Cross-entropy is the natural choice for neural network classification because it's the negative log-likelihood of the correct class, making it theoretically grounded in maximum likelihood estimation. It's also convex with respect to the final layer's logits, ensuring stable gradient descent optimization. 

### Overfitting and Underfitting Prevention

#### 1. Overfitting Prevention

1. Frozen Pretrained Embeddings: GloVe embeddings are pretrained on billions of words from Wikipedia and news corpora. These vectors already encode rich semantic relationships. By freezing these weights, we prevent the model from overfitting embeddings to our small training set (6,568 samples). 


2. Dropout Regularization: Dropout randomly sets activations to zero during training with probability `p`, forcing the network to learn redundant representations. If the model relies on a single neuron to detect "not," and that neuron is dropped out 50% of the time, the model must learn alternative pathways to detect negation.

3. Learning Rate Scheduling: Learning rate scheduling automatically adjusts the learning rate during training to balance fast learning with precise convergence. Early in training, a high learning rate (0.15) allows the model to quickly explore the parameter space and make rapid progress toward good solutions. However, as training progresses, these large update steps cause the model to overshoot optimal parameter values, bouncing around the best solution rather than settling into it. The scheduler monitors validation loss and reduces the learning rate by 10% (factor=0.9) when performance plateaus for 3 consecutive epochs.

4. Early Stopping with Model Checkpointing

Training for too many epochs inevitably leads to overfitting—the model continues improving on training data while validation performance degrades. Early stopping prevents this by monitoring validation loss and saving the model checkpoint when it achieves the lowest validation loss. After 50 epochs, even if the final model is overfit, we retain the best-performing model from earlier in training (typically around epochs 15-25).

5. Train-Validation Monitoring

Explicit monitoring of the train-validation gap provides a human-interpretable diagnostic for model health. A gap of 2-5% is normal and acceptable—it indicates the model has learned patterns specific to training data but still generalizes well. A gap exceeding 10% signals severe overfitting, prompting us to stop training, increase regularization, or reduce model capacity.

### 2. Underfitting Prevention

1. Sufficient Model Capacity: 2-layer BiLSTM with 384 hidden units provides `384 × 2 directions × 2 layers = 1,536` hidden states across the network, sufficient to learn nuanced linguistic patterns.

2. Adequate Training Duration: Training for 50 epochs with early stopping ensures we explore enough of the loss landscape. If validation loss is still decreasing at epoch 50, we'd extend training, but typically loss plateaus around epoch 20-30.

3. High Initial Learning Rate: Starting with LR=0.15 ensures rapid initial learning. Too small a learning rate (e.g., 0.001) would cause extremely slow convergence, potentially getting stuck in poor local minima and resulting in underfitting.

4. Monitoring Training Loss: If training loss remains high (>0.5) throughout training, this signals underfitting—the model lacks capacity or training time to fit the data. We'd respond by adding layers, or training longer.

In [ ]:
hidden_dim = 384  # LSTM hidden state dimension
label_size = 2    # Binary classification: positive (1) vs negative (0)
num_selected_features = len(selected_features)  # Number of hand-crafted features to use

# Display model architecture configuration for reproducibility
print(f"Model configuration:")
print(f"  - Vocabulary size: {vocab_size}")           # Total unique tokens
print(f"  - Embedding dimension: {embedding_dim}")     # GloVe vector size (100)
print(f"  - LSTM hidden dimension: {hidden_dim}")      # Hidden state size
print(f"  - Number of features used: {num_selected_features}")  # Selected features


class BiLSTMClassifier(nn.Module):
    """
    Bidirectional LSTM sentiment classifier with hand-crafted feature fusion.
    
    Architecture:
        1. Embedding Layer: Maps token IDs to GloVe vectors (frozen)
        2. BiLSTM Layers: Captures bidirectional sequential context (2 layers, dropout 0.5)
        3. Feature Network: Projects hand-crafted features to dense representation
        4. Fusion Layer: Concatenates LSTM output with feature vector
        5. Classification Head: Final fully-connected layer for sentiment prediction
    """
    
    def __init__(self, vocab_size, embedding_dim, hidden_dim, label_size, 
                 padding_idx, num_features, selected_feat_names):
        """
        Initialize the hybrid BiLSTM classifier.
        """
        super(BiLSTMClassifier, self).__init__()
        
        # Embedding layer, create embedding lookup table: vocab_size x embedding_dim
        self.embedding = nn.Embedding(vocab_size, embedding_dim, padding_idx=padding_idx)
        
        # Load pretrained GloVe vectors into embedding weights
        self.embedding.weight.data.copy_(TEXT.vocab.vectors)
        
        # Don't update during training
        self.embedding.weight.requires_grad = False
        
        # Bidirectuinal LSTM (2 layers, dropout 0.5)
        # - Forward LSTM: reads text left-to-right (past → present)
        # - Backward LSTM: reads text right-to-left (future → present)
        self.lstm = nn.LSTM(
            embedding_dim,      # Input size: word embedding dimension (100)
            hidden_dim,         # Hidden state size: 384 per direction
            num_layers=2,       # Stack 2 LSTM layers for deeper representation
            batch_first=True,   # Input shape: (batch, seq_len, features)
            dropout=0.5,        # Dropout between LSTM layers (regularization)
            bidirectional=True  # Process sequence in both directions
        )
        
        # Hand-crafted feature projection network
        self.numeric_proj = nn.Sequential(
            nn.Linear(num_features, 32),  # Project features to 32-dim space
            nn.ReLU(),                     # Non-linear activation
            nn.Dropout(0.2)                # Light dropout for regularization
        )
        
        # Fusion + Classification head
        # Fuse LSTM representation (768-dim) with feature vector (32-dim)
        self.fc = nn.Linear(hidden_dim * 2 + 32, label_size)
        
        # Dropout for embedding layer (applied after embedding lookup)
        self.dropout = nn.Dropout(0.5)
        
        # Store padding index and feature names for forward pass
        self.pad_idx = padding_idx
        self.selected_feat_names = selected_feat_names

    def forward(self, x):
        """
        Forward pass: Convert token IDs to sentiment predictions.
        """
        # Embed tokens (with dropout regularization)
        embedded = self.dropout(self.embedding(x))
        
        # Process sequence through bidirectional LSTM
        # We only need the final hidden states from both directions
        _, (hidden, _) = self.lstm(embedded)
        
        # Extract final layer's hidden states from both directions
        hidden = torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim=1)
        
        # Apply dropout to LSTM output for regularization
        sent_vec = self.dropout(hidden)
        
        # Compute linguistic features from token IDs
        numeric_feats = compute_numeric_features(x, self.pad_idx, self.selected_feat_names)
        
        # Project features through small neural network
        numeric_vec = self.numeric_proj(numeric_feats)
        
        # Concatenate LSTM output (768-dim) with feature projection (32-dim)
        fused = torch.cat([sent_vec, numeric_vec], dim=1)
    
        # Map fused representation to class logits
        return self.fc(fused)


# Create the hybrid BiLSTM classifier with all specified parameters
model = BiLSTMClassifier(
    vocab_size,              # Total vocabulary size
    embedding_dim,           # GloVe embedding dimension (100)
    hidden_dim,              # LSTM hidden size (384)
    label_size,              # Binary classification (2)
    padding_idx,             # Padding token index
    num_selected_features,   # Number of hand-crafted features
    selected_features        # List of feature names to extract
)

# Move model to GPU if available (enables parallel computation)
model.to(device)


# Loss function: Cross-entropy for multi-class classification
criterion = nn.CrossEntropyLoss()

# Optimizer: Stochastic Gradient Descent with momentum
# Only optimize parameters that require gradients (excludes frozen embeddings)
optimizer = optim.SGD(
    filter(lambda p: p.requires_grad, model.parameters()),  # Only trainable params
    lr=0.15,        # Learning rate (relatively high for SGD)
    momentum=0.9    # Momentum helps escape local minima and smooths updates
)

# Learning rate scheduler: Reduce LR when validation loss plateaus
scheduler = optim.lr_scheduler.ReduceLROnPlateau(
    optimizer,
    'min',          # Monitor validation loss (minimize)
    factor=0.9,     # Multiply LR by 0.9 when plateau detected
    patience=3      # Wait 3 epochs before reducing LR
)

# Display overfitting prevention strategies for transparency
print("Overfitting prevention strategies:")
print("  1. Freeze pretrained embeddings (reduce parameters)")
print("  2. Dropout (0.5 LSTM, 0.2 feature layer)")
print("  3. Learning Rate Scheduler (reduce LR when validation stalls)")
print("  4. Early Stopping (save best validation model)")
print("  5. Monitor train/validation gap")


def train(model, iterator, optimizer, criterion):
    """
    Train the model for one epoch.
    """
    # Set model to training mode
    model.train()
    
    epoch_loss = 0  # Accumulator for batch losses
    
    # Iterate through all training batches
    for batch in iterator:
        # Zero out gradients from previous batch
        optimizer.zero_grad()
        
        # Move batch data to GPU
        x = batch.text.to(device, non_blocking=True)   # Input token IDs
        y = batch.label.to(device, non_blocking=True)  # Ground truth labels
        
        # Compute predictions
        pred = model(x) 
        
        # Compute loss
        loss = criterion(pred, y)  # Cross-entropy loss (scalar)
        
        # Compute gradients via backpropagation
        loss.backward()
        
        # Update weights using gradients
        optimizer.step()  # weights = weights - lr * gradients
        
        # Accumulate loss for epoch average
        epoch_loss += loss.item()
    
    # Return average loss across all batches
    return epoch_loss / len(iterator)

def evaluate(model, iterator, criterion):
    """
    Evaluate the model on validation/test set.
    """
    # Set model to evaluation mode
    model.eval()
    
    epoch_loss = 0
    
    # Disable gradient computation for efficiency
    with torch.no_grad():
        for batch in iterator:
            # Move data to GPU
            x = batch.text.to(device, non_blocking=True)
            y = batch.label.to(device, non_blocking=True)
            
            # Forward pass only (no backward pass needed)
            pred = model(x)
            loss = criterion(pred, y)
            
            epoch_loss += loss.item()
    
    return epoch_loss / len(iterator)


def accuracy(model, iterator):
    """
    Compute classification accuracy on a dataset.
    """
    model.eval()
    correct = 0  # Count of correct predictions
    total = 0    # Total number of samples
    
    with torch.no_grad():
        for batch in iterator:
            x = batch.text.to(device, non_blocking=True)
            y = batch.label.to(device, non_blocking=True)
            
            # Get model predictions (logits)
            pred = model(x)
            
            # Convert logits to class predictions
            _, predicted = torch.max(pred, 1)  # Get class with highest score
            
            # Count total samples and correct predictions
            total += y.size(0)
            correct += (predicted == y).sum().item()
    
    # Return accuracy as percentage (0.0 to 1.0)
    return correct / total

N_EPOCHS = 50  # Maximum number of training epochs
best_valid_loss = float('inf')  # Track best validation loss for early stopping

# Lists to store metrics for plotting learning curves
train_losses, val_losses = [], []  # Loss trajectories
train_accs, val_accs = [], []      # Accuracy trajectories

print(f"\nStarting training for {N_EPOCHS} epochs...")

# Main training loop
for epoch in range(N_EPOCHS):
    # Train for one epoch
    train_loss = train(model, train_iter, optimizer, criterion)
    
    # Evaluate on validation set
    valid_loss = evaluate(model, val_iter, criterion)
    
    # Comnpute accuracies
    train_acc = accuracy(model, train_iter)
    valid_acc = accuracy(model, val_iter)
    
    # Record metrics for analysis
    train_losses.append(train_loss)
    val_losses.append(valid_loss)
    train_accs.append(train_acc)
    val_accs.append(valid_acc)
    
    # Reduce learning rate if validation loss hasn't improved
    scheduler.step(valid_loss)
    
    # Keep track of model with lowest validation loss
    if valid_loss < best_valid_loss:
        best_valid_loss = valid_loss
        best_model = copy.deepcopy(model)  # Deep copy to preserve weights
    
    # Periodic logging every 5 epochs
    if (epoch + 1) % 5 == 0:
        print(f'Epoch: {epoch+1:02} | LR: {scheduler.get_last_lr()[-1]:.4f}')
        print(f'  Train Loss: {train_loss:.3f} | Train Acc: {train_acc:.3f}')
        print(f'  Val Loss: {valid_loss:.3f} | Val Acc: {valid_acc:.3f}')
        
        # Monitor overfitting: Large gap between train/val indicates overfitting
        # If train_acc >> val_acc, model is memorizing rather than generalizing
        print(f'  Overfitting check: Train/Val Acc gap = {abs(train_acc - valid_acc):.3f}')

In [ ]:
test_acc = accuracy(best_model, test_iter)

print("\nPerformance Comparison:")
print(f"  Logistic Regression Baseline:  {lr_acc:.4f}")
print(f"  BiLSTM + Feature Fusion (val): {max(val_accs):.4f}")
print(f"  BiLSTM + Feature Fusion (test):{test_acc:.4f}")
print(f"  Performance improvement:        +{max(val_accs) - lr_acc:.4f}")

print("Summary:")
print(f"  1. Feature Engineering: Selected {num_selected_features} from 15 candidate features")
print(f"  2. Feature Validation: LR baseline {lr_acc:.4f}")
print(f"  3. Deep Model: BiLSTM with feature fusion achieved {test_acc:.4f}")
print(f"  4. Explainability: Chi-square test + Ablation study validated feature effectiveness")